## ДЗ 1 (ОБЯЗАТЕЛЬНОЕ): Анализ температурных данных и мониторинг текущей температуры через OpenWeatherMap API

**Описание задания:**  
Вы аналитик в компании, занимающейся изучением климатических изменений и мониторингом температур в разных городах. Вам нужно провести анализ исторических данных о температуре для выявления сезонных закономерностей и аномалий. Также необходимо подключить API OpenWeatherMap для получения текущей температуры в выбранных городах и сравнить её с историческими данными.


### Цели задания:
1. Провести **анализ временных рядов**, включая:
   - Вычисление скользящего среднего и стандартного отклонения для сглаживания температурных колебаний.
   - Определение аномалий на основе отклонений температуры от $ \text{скользящее среднее} \pm 2\sigma $.
   - Построение долгосрочных трендов изменения температуры.
   - Любые дополнительные исследования будут вам в плюс.

2. Осуществить **мониторинг текущей температуры**:
   - Получить текущую температуру через OpenWeatherMap API.
   - Сравнить её с историческим нормальным диапазоном для текущего сезона.

3. Разработать **интерактивное приложение**:
   - Дать пользователю возможность выбрать город.
   - Отобразить результаты анализа температур, включая временные ряды, сезонные профили и аномалии.
   - Провести анализ текущей температуры в контексте исторических данных.


### Описание данных
Исторические данные о температуре содержатся в файле `temperature_data.csv`, включают:
  - `city`: Название города.
  - `timestamp`: Дата (с шагом в 1 день).
  - `temperature`: Среднесуточная температура (в °C).
  - `season`: Сезон года (зима, весна, лето, осень).

Код для генерации файла вы найдете ниже.

### Этапы выполнения

1. **Анализ исторических данных**:
   - Вычислить **скользящее среднее** температуры с окном в 30 дней для сглаживания краткосрочных колебаний.
   - Рассчитать среднюю температуру и стандартное отклонение для каждого сезона в каждом городе.
   - Выявить аномалии, где температура выходит за пределы $ \text{среднее} \pm 2\sigma $.
   - Попробуйте распараллелить проведение этого анализа. Сравните скорость выполнения анализа с распараллеливанием и без него.

2. **Мониторинг текущей температуры**:
   - Подключить OpenWeatherMap API для получения текущей температуры города. Для получения API Key (бесплатно) надо зарегистрироваться на сайте. Обратите внимание, что API Key может активироваться только через 2-3 часа, это нормально. Посему получите ключ заранее.
   - Получить текущую температуру для выбранного города через OpenWeatherMap API.
   - Определить, является ли текущая температура нормальной, исходя из исторических данных для текущего сезона.
   - Данные на самом деле не совсем реальные (сюрпрайз). Поэтому на момент эксперимента погода в Берлине, Каире и Дубае была в рамках нормы, а в Пекине и Москве аномальная. Протестируйте свое решение для разных городов.
   - Попробуйте для получения текущей температуры использовать синхронные и асинхронные методы. Что здесь лучше использовать?

3. **Создание приложения на Streamlit**:
   - Добавить интерфейс для загрузки файла с историческими данными.
   - Добавить интерфейс для выбора города (из выпадающего списка).
   - Добавить форму для ввода API-ключа OpenWeatherMap. Когда он не введен, данные для текущей погоды не показываются. Если ключ некорректный, выведите на экран ошибку (должно приходить `{"cod":401, "message": "Invalid API key. Please see https://openweathermap.org/faq#error401 for more info."}`).
   - Отобразить:
     - Описательную статистику по историческим данным для города, можно добавить визуализации.
     - Временной ряд температур с выделением аномалий (например, точками другого цвета).
     - Сезонные профили с указанием среднего и стандартного отклонения.
   - Вывести текущую температуру через API и указать, нормальна ли она для сезона.

### Критерии оценивания

- Корректное проведение анализа данных – 1 балл.
- Исследование распараллеливания анализа – 1 балл.
- Корректный поиск аномалий – 1 балл.
- Подключение к API и корректность выполнения запроса – 1 балл.
- Проведение эксперимента с синхронным и асинхронным способом запроса к API – 1 балл.
- Создание интерфейса приложения streamlit в соответствии с описанием – 3 балла.
- Корректное отображение графиков и статистик, а также сезонных профилей – 1 балл.
- Корректный вывод текущей температуры в выбранном городе и проведение проверки на ее аномальность – 1 балл.
- Любая дополнительная функциональность приветствуется и оценивается бонусными баллами (не более 2 в сумме) на усмотрение проверяющего.

### Формат сдачи домашнего задания

Решение нужно развернуть в Streamlit Cloud (бесплатно)

*   Создаем новый репозиторий на GitHub.  
*   Загружаем проект.
*   Создаем аккаунт в [Streamlit Cloud](https://streamlit.io/cloud).
*   Авторизуемся в Streamlit Cloud.
*   Создаем новое приложение в Streamlit Cloud и подключаем GitHub-репозиторий.
*   Deploy!

Сдать в форму необходимо:
1. Ссылку на развернутое в Streamlit Cloud приложение.
2. Ссылку на код. Все выводы про, например, использование параллельности/асинхронности опишите в комментариях.

Не забудьте удалить ключ API и иную чувствительную информацию.

### Полезные ссылки
*   [Оформление задачи Титаник на Streamlit](https://github.com/evgpat/streamlit_demo)
*   [Документация Streamlit](https://docs.streamlit.io/)
*   [Блог о Streamlit](https://blog.streamlit.io/)

In [97]:
import pandas as pd
import numpy as np
import time
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import OneHotEncoder
from multiprocessing import Process, Manager
import aiohttp
import asyncio
import nest_asyncio

# Реальные средние температуры (примерные данные) для городов по сезонам
seasonal_temperatures = {
    "New York": {"winter": 0, "spring": 10, "summer": 25, "autumn": 15},
    "London": {"winter": 5, "spring": 11, "summer": 18, "autumn": 12},
    "Paris": {"winter": 4, "spring": 12, "summer": 20, "autumn": 13},
    "Tokyo": {"winter": 6, "spring": 15, "summer": 27, "autumn": 18},
    "Moscow": {"winter": -10, "spring": 5, "summer": 18, "autumn": 8},
    "Sydney": {"winter": 12, "spring": 18, "summer": 25, "autumn": 20},
    "Berlin": {"winter": 0, "spring": 10, "summer": 20, "autumn": 11},
    "Beijing": {"winter": -2, "spring": 13, "summer": 27, "autumn": 16},
    "Rio de Janeiro": {"winter": 20, "spring": 25, "summer": 30, "autumn": 25},
    "Dubai": {"winter": 20, "spring": 30, "summer": 40, "autumn": 30},
    "Los Angeles": {"winter": 15, "spring": 18, "summer": 25, "autumn": 20},
    "Singapore": {"winter": 27, "spring": 28, "summer": 28, "autumn": 27},
    "Mumbai": {"winter": 25, "spring": 30, "summer": 35, "autumn": 30},
    "Cairo": {"winter": 15, "spring": 25, "summer": 35, "autumn": 25},
    "Mexico City": {"winter": 12, "spring": 18, "summer": 20, "autumn": 15},
}

# Сопоставление месяцев с сезонами
month_to_season = {12: "winter", 1: "winter", 2: "winter",
                   3: "spring", 4: "spring", 5: "spring",
                   6: "summer", 7: "summer", 8: "summer",
                   9: "autumn", 10: "autumn", 11: "autumn"}

# Генерация данных о температуре
def generate_realistic_temperature_data(cities, num_years=10):
    dates = pd.date_range(start="2010-01-01", periods=365 * num_years, freq="D")
    data = []

    for city in cities:
        for date in dates:
            season = month_to_season[date.month]
            mean_temp = seasonal_temperatures[city][season]
            # Добавляем случайное отклонение
            temperature = np.random.normal(loc=mean_temp, scale=5)
            data.append({"city": city, "timestamp": date, "temperature": temperature})

    df = pd.DataFrame(data)
    df['season'] = df['timestamp'].dt.month.map(lambda x: month_to_season[x])
    return df

# Генерация данных
data = generate_realistic_temperature_data(list(seasonal_temperatures.keys()))
data.to_csv('temperature_data.csv', index=False)


## Часть 1. Анализ исторических данных

Посмотрим на данные:

In [98]:
data.sample(5)

city  timestamp  temperature  season
13404     Tokyo 2016-09-20    11.307278  autumn
3623   New York 2019-12-03    -3.112461  winter
26651   Beijing 2013-01-06     2.471301  winter
24337    Berlin 2016-09-03    15.189800  autumn
1280   New York 2013-07-04    28.045656  summer

Посмотрим на типы данных:

In [99]:
data.dtypes

city                   object
timestamp      datetime64[ns]
temperature           float64
season                 object
dtype: object

Посмотрим на пропуски:

In [100]:
data.isna().sum()

city           0
timestamp      0
temperature    0
season         0
dtype: int64

Пропусков - нет.

Посмотрим на дубли:

In [101]:
data.duplicated().sum()

0

Дублей - нет.

Посмотрим на уникальные значения колонок `city` и `season`:

In [102]:
data['city'].unique()

array(['New York', 'London', 'Paris', 'Tokyo', 'Moscow', 'Sydney',
       'Berlin', 'Beijing', 'Rio de Janeiro', 'Dubai', 'Los Angeles',
       'Singapore', 'Mumbai', 'Cairo', 'Mexico City'], dtype=object)

In [103]:
data['season'].unique()

array(['winter', 'spring', 'summer', 'autumn'], dtype=object)

Посмотрим на размах дат в колонке `timestamp`:

In [104]:
data['timestamp'].min(), data['timestamp'].max()

(Timestamp('2010-01-01 00:00:00'), Timestamp('2019-12-29 00:00:00'))

__1. Вычислить скользящее среднее температуры с окном в 30 дней для сглаживания краткосрочных колебаний:__

Давайте сперва отсортируем наши данные:

In [105]:
data = data.sort_values(by=['city', 'timestamp'], ascending=[True, True])

In [106]:
data['rolling_mean_30d'] = data.groupby(['city', 'season'])['temperature'].transform(lambda row: row.rolling(window=30, min_periods=1).mean())

In [107]:
data.head(5)

city  timestamp  temperature  season  rolling_mean_30d
25550  Beijing 2010-01-01     2.893672  winter          2.893672
25551  Beijing 2010-01-02     0.214542  winter          1.554107
25552  Beijing 2010-01-03    -6.675774  winter         -1.189186
25553  Beijing 2010-01-04    -3.568574  winter         -1.784033
25554  Beijing 2010-01-05    -1.460492  winter         -1.719325

__2. Рассчитать среднюю температуру и стандартное отклонение для каждого сезона в каждом городе:__

In [108]:
data['avg_temp'] = data.groupby(['city', 'season'])['temperature'].transform(lambda row: row.mean())
data['std_temp'] = data.groupby(['city', 'season'])['temperature'].transform(lambda row: row.std())

In [109]:
data.head(5)

city  timestamp  temperature  season  rolling_mean_30d  avg_temp  \
25550  Beijing 2010-01-01     2.893672  winter          2.893672 -1.925424   
25551  Beijing 2010-01-02     0.214542  winter          1.554107 -1.925424   
25552  Beijing 2010-01-03    -6.675774  winter         -1.189186 -1.925424   
25553  Beijing 2010-01-04    -3.568574  winter         -1.784033 -1.925424   
25554  Beijing 2010-01-05    -1.460492  winter         -1.719325 -1.925424   

       std_temp  
25550  5.103989  
25551  5.103989  
25552  5.103989  
25553  5.103989  
25554  5.103989

__3. Выявить аномалии, где температура выходит за пределы среднее±2σ:__

Вычислим нижнюю и верхнюю границы, которые будут считаться как средняя температура ±2σ:

In [110]:
data['lower_bound'] = data['avg_temp'] - 2 * data['std_temp']
data['upper_bound'] = data['avg_temp'] + 2 * data['std_temp']

In [111]:
data.head(5)

city  timestamp  temperature  season  rolling_mean_30d  avg_temp  \
25550  Beijing 2010-01-01     2.893672  winter          2.893672 -1.925424   
25551  Beijing 2010-01-02     0.214542  winter          1.554107 -1.925424   
25552  Beijing 2010-01-03    -6.675774  winter         -1.189186 -1.925424   
25553  Beijing 2010-01-04    -3.568574  winter         -1.784033 -1.925424   
25554  Beijing 2010-01-05    -1.460492  winter         -1.719325 -1.925424   

       std_temp  lower_bound  upper_bound  
25550  5.103989   -12.133401     8.282554  
25551  5.103989   -12.133401     8.282554  
25552  5.103989   -12.133401     8.282554  
25553  5.103989   -12.133401     8.282554  
25554  5.103989   -12.133401     8.282554

Давайте создадим техническую колонку `is_anomaly`, которая будет принимать значение 1, если temperature > upper_bound или temperature < lower_bound, и 0 - в противном случае:

In [112]:
data['is_anomaly'] = np.where((data['temperature'] > data['upper_bound'])|(data['temperature'] < data['lower_bound']), 1, 0)
data.head(5)

city  timestamp  temperature  season  rolling_mean_30d  avg_temp  \
25550  Beijing 2010-01-01     2.893672  winter          2.893672 -1.925424   
25551  Beijing 2010-01-02     0.214542  winter          1.554107 -1.925424   
25552  Beijing 2010-01-03    -6.675774  winter         -1.189186 -1.925424   
25553  Beijing 2010-01-04    -3.568574  winter         -1.784033 -1.925424   
25554  Beijing 2010-01-05    -1.460492  winter         -1.719325 -1.925424   

       std_temp  lower_bound  upper_bound  is_anomaly  
25550  5.103989   -12.133401     8.282554           0  
25551  5.103989   -12.133401     8.282554           0  
25552  5.103989   -12.133401     8.282554           0  
25553  5.103989   -12.133401     8.282554           0  
25554  5.103989   -12.133401     8.282554           0

Давайте выведем записи, где `is_anomaly` равен 1:

In [113]:
data[data['is_anomaly'] == 1]

city  timestamp  temperature  season  rolling_mean_30d   avg_temp  \
25612  Beijing 2010-03-04    24.090868  spring         16.820474  12.935693   
25633  Beijing 2010-03-25    -0.229748  spring         11.444821  12.935693   
25643  Beijing 2010-04-04    25.000456  spring         10.720768  12.935693   
25734  Beijing 2010-07-04    40.059700  summer         27.727083  27.075394   
25739  Beijing 2010-07-09    37.159025  summer         28.245132  27.075394   
...        ...        ...          ...     ...               ...        ...   
14500    Tokyo 2019-09-21     4.002044  autumn         17.407992  17.784407   
14555    Tokyo 2019-11-15     4.881912  autumn         17.224884  17.784407   
14560    Tokyo 2019-11-20     2.256648  autumn         17.017363  17.784407   
14565    Tokyo 2019-11-25     5.617291  autumn         16.483542  17.784407   
14585    Tokyo 2019-12-15    -4.108986  winter          6.564711   6.046550   

       std_temp  lower_bound  upper_bound  is_anomaly  
25612  5.136070     2.663552    23.207834           1  
25633  5.136070     2.663552    23.207834           1  
25643  5.136070     2.663552    23.207834           1  
25734  4.902326    17.270742    36.880046           1  
25739  4.902326    17.270742    36.880046           1  
...         ...          ...          ...         ...  
14500  5.271330     7.241746    28.327068           1  
14555  5.271330     7.241746    28.327068           1  
14560  5.271330     7.241746    28.327068           1  
14565  5.271330     7.241746    28.327068           1  
14585  4.892767    -3.738983    15.832083           1  

[2441 rows x 10 columns]

__4. Построение долгосрочных трендов изменения температуры:__

Давайте построим долгосрочные тренды изменения температуры с помощью линейной регрессии от дня недели и времени года:

In [114]:
data['day_of_week'] = data['timestamp'].dt.dayofweek

Закодируем день недели, сезон и город:

In [115]:
encoder = OneHotEncoder(sparse_output=False, drop='first')
encoded_features = encoder.fit_transform(data[['day_of_week', 'season', 'city']])

In [116]:
encoded_df = pd.DataFrame(encoded_features, columns=encoder.get_feature_names_out(['day_of_week', 'season', 'city']))

In [117]:
encoded_df.head(5)

day_of_week_1  day_of_week_2  day_of_week_3  day_of_week_4  day_of_week_5  \
0            0.0            0.0            0.0            1.0            0.0   
1            0.0            0.0            0.0            0.0            1.0   
2            0.0            0.0            0.0            0.0            0.0   
3            0.0            0.0            0.0            0.0            0.0   
4            1.0            0.0            0.0            0.0            0.0   

   day_of_week_6  season_spring  season_summer  season_winter  city_Berlin  \
0            0.0            0.0            0.0            1.0          0.0   
1            0.0            0.0            0.0            1.0          0.0   
2            1.0            0.0            0.0            1.0          0.0   
3            0.0            0.0            0.0            1.0          0.0   
4            0.0            0.0            0.0            1.0          0.0   

   ...  city_Los Angeles  city_Mexico City  city_Moscow  city_Mumbai  \
0  ...               0.0               0.0          0.0          0.0   
1  ...               0.0               0.0          0.0          0.0   
2  ...               0.0               0.0          0.0          0.0   
3  ...               0.0               0.0          0.0          0.0   
4  ...               0.0               0.0          0.0          0.0   

   city_New York  city_Paris  city_Rio de Janeiro  city_Singapore  \
0            0.0         0.0                  0.0             0.0   
1            0.0         0.0                  0.0             0.0   
2            0.0         0.0                  0.0             0.0   
3            0.0         0.0                  0.0             0.0   
4            0.0         0.0                  0.0             0.0   

   city_Sydney  city_Tokyo  
0          0.0         0.0  
1          0.0         0.0  
2          0.0         0.0  
3          0.0         0.0  
4          0.0         0.0  

[5 rows x 23 columns]

Добавим закодированные данные:

In [118]:
data = pd.concat([data, encoded_df], axis=1)

Разделим данные на матрицу объект-признак с признаками и вектор столбец с целевой переменной:

In [119]:
data.columns

Index(['city', 'timestamp', 'temperature', 'season', 'rolling_mean_30d',
       'avg_temp', 'std_temp', 'lower_bound', 'upper_bound', 'is_anomaly',
       'day_of_week', 'day_of_week_1', 'day_of_week_2', 'day_of_week_3',
       'day_of_week_4', 'day_of_week_5', 'day_of_week_6', 'season_spring',
       'season_summer', 'season_winter', 'city_Berlin', 'city_Cairo',
       'city_Dubai', 'city_London', 'city_Los Angeles', 'city_Mexico City',
       'city_Moscow', 'city_Mumbai', 'city_New York', 'city_Paris',
       'city_Rio de Janeiro', 'city_Singapore', 'city_Sydney', 'city_Tokyo'],
      dtype='object')

In [120]:
X = data.drop(['timestamp', 'city', 'temperature', 'lower_bound', 'upper_bound', 'is_anomaly', 'season'], axis=1)
y = data['temperature']

Обучим модель:

In [121]:
model = LinearRegression()
model.fit(X, y)

LinearRegression()

Сделаем прогноз:

In [122]:
data['predicted_temp'] = model.predict(X)

In [123]:
data['predicted_temp'][:5]

25550    2.803290
25551    1.488657
25552   -1.218022
25553   -1.689473
25554   -1.745117
Name: predicted_temp, dtype: float64

__5. Попробуйте распараллелить проведение этого анализа. Сравните скорость выполнения анализа с распараллеливанием и без него:__

Давайте сперва посчитаем время последовательного выполнения кода без распараллеливания, для этого оформим код анализа в виде функции:

In [124]:
# Генерация данных
data_for_speed_test =  generate_realistic_temperature_data(list(seasonal_temperatures.keys()))

In [125]:
def get_weather_statistics(city_name, df):
    city_data = df[df['city'] == city_name]
    city_data.sort_values(by='timestamp', ascending=True, inplace=True)

    city_data['rolling_mean_30d'] = city_data.groupby('season')['temperature'].transform(lambda row: row.rolling(window=30, min_periods=1).mean())

    city_data['avg_temp'] = city_data.groupby('season')['temperature'].transform(lambda row: row.mean())
    city_data['std_temp'] = city_data.groupby('season')['temperature'].transform(lambda row: row.std())

    city_data['lower_bound'] = city_data['avg_temp'] - 2 * data['std_temp']
    city_data['upper_bound'] = city_data['avg_temp'] + 2 * data['std_temp']

    city_data['is_anomaly'] = np.where((city_data['temperature'] > city_data['upper_bound'])|(city_data['temperature'] < city_data['lower_bound']), 1, 0)
    return city_data

In [126]:
start_time = time.time()
get_weather_statistics('Moscow', data_for_speed_test)
end_time = time.time()
print(f"Время выполнения без распараллеливания составляет: {end_time - start_time:.2f} секунд")

Время выполнения без распараллеливания составляет: 0.08 секунд


<ipython-input-125-696e0acbe967>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  city_data.sort_values(by='timestamp', ascending=True, inplace=True)
<ipython-input-125-696e0acbe967>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  city_data['rolling_mean_30d'] = city_data.groupby('season')['temperature'].transform(lambda row: row.rolling(window=30, min_periods=1).mean())
<ipython-input-125-696e0acbe967>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the c

Теперь давайте посчитаем время выполнения кода с распаралеливанием:

In [31]:
# Генерация данных
data_for_speed_test =  generate_realistic_temperature_data(list(seasonal_temperatures.keys()))

Попробуем обрабатывать данные для отдельных сезонов и сохранять результат в словарь:

In [127]:
def parallel_func(season_df, results, season):
    season_df = season_df.sort_values(by='timestamp', ascending=True)

    # Вычисления для rolling mean, avg_temp, std_temp, границ и аномалий
    season_df['rolling_mean_30d'] = season_df['temperature'].rolling(window=30, min_periods=1).mean()

    avg_temp = season_df['temperature'].mean()
    std_temp = season_df['temperature'].std()

    season_df['lower_bound'] = avg_temp - 2 * std_temp
    season_df['upper_bound'] = avg_temp + 2 * std_temp

    season_df['is_anomaly'] = np.where((season_df['temperature'] > season_df['upper_bound'])|(season_df['temperature'] < season_df['lower_bound']), 1, 0)

    results[season] = season_df

def get_weather_statistics_parallel(city_name, df):
    df = df[df['city'] == city_name]
    grouped = df.groupby('season')

    manager = Manager()
    results = manager.dict()
    processes = []

    for season, group in grouped:
        process = Process(target=parallel_func, args=(group, results, season))
        processes.append(process)
        process.start()

    for process in processes:
        process.join()

    return pd.concat(list(results.values()))

start_time = time.time()
result = get_weather_statistics_parallel('Moscow', data_for_speed_test)
end_time = time.time()
print(f"Время выполнения c распараллеливанием составляет: {end_time - start_time:.2f} секунд")

Время выполнения c распараллеливанием составляет: 0.33 секунд


Последовательное выполнение кода без распараллеливания оказалось быстрее: 0.08 cекунд против 0.33 секунд.

## Часть 2. Мониторинг текущей температуры

__1. Подключить OpenWeatherMap API для получения текущей температуры города. Для получения API Key (бесплатно) надо зарегистрироваться на сайте. Обратите внимание, что API Key может активироваться только через 2-3 часа, это нормально. Посему получите ключ заранее:__

API Key получен.

__2. Получить текущую температуру для выбранного города через OpenWeatherMap API:__

In [128]:
api_key = "MY_API_KEY"

def get_current_temperature(city_name):
    url = "https://api.openweathermap.org/data/2.5/weather"
    params = {"q": city_name,
              "appid": api_key,
              "units": "metric"}
    try:
        response = requests.get(url, params=params)
        #print(response)

        data = response.json()
        temperature = data["main"]["temp"]
        #print(temperature)
        time = data["sys"]["sunrise"]
        season = pd.to_datetime(time, unit='s').month
        #print(time)
        return f"Текущая температура в городе {city_name}: {temperature}°C", temperature, season
    except requests.exceptions.RequestException as e:
        return f"Ошибка при выполнении запроса: {e}"
    except KeyError:
        return "Ошибка: не удается получить информацию о температуре"

In [129]:
data['city'].unique()

array(['Beijing', 'Berlin', 'Cairo', 'Dubai', 'London', 'Los Angeles',
       'Mexico City', 'Moscow', 'Mumbai', 'New York', 'Paris',
       'Rio de Janeiro', 'Singapore', 'Sydney', 'Tokyo'], dtype=object)

Выведем темперутуру для Москвы:

In [130]:
city_name = input("Введите название города: ")
print(get_current_temperature(city_name)[0])

Введите название города: Moscow
Текущая температура в городе Moscow: 1.55°C


__3. Определить, является ли текущая температура нормальной, исходя из исторических данных для текущего сезона:__

Выведем значение текущей температуры:

In [131]:
city_name

'Moscow'

In [132]:
current_temp = get_current_temperature(city_name)[1]
current_temp

1.55

In [133]:
def is_temperature_normal(city, temperature, month, df):
    df = df[(df["city"] == city) & (df["timestamp"].dt.month == month)]

    lower_bound = df.iloc[0]["lower_bound"]
    upper_bound = df.iloc[0]["upper_bound"]

    if lower_bound <= temperature <= upper_bound:
        return 'Температура в норме'
    else:
        return 'Температура выходит за пределы нормы'

In [134]:
is_temperature_normal(city_name, current_temp, get_current_temperature(city_name)[2], data)

'Температура выходит за пределы нормы'

__4. Данные на самом деле не совсем реальные (сюрпрайз). Поэтому на момент эксперимента погода в Берлине, Каире и Дубае была в рамках нормы, а в Пекине и Москве аномальная. Протестируйте свое решение для разных городов:__

Давайте проверим теперь на аномальность температуры в городах Берлин, Каир, Думбае:

Посмотрим для города Берлин (Berlin)

In [135]:
city_name = input("Введите название города: ")
current_temp = get_current_temperature(city_name)[1]
month = get_current_temperature(city_name)[2]
is_temperature_normal(city_name, current_temp, month, data)

Введите название города: Berlin


'Температура в норме'

Посмотрим для города Каир (Cairo):

In [45]:
city_name = input("Введите название города: ")
current_temp = get_current_temperature(city_name)[1]
month = get_current_temperature(city_name)[2]
is_temperature_normal(city_name, current_temp, month, data)

Введите название города: Cairo


'Температура в норме'

Посмотрим для города Дубай (Dubai):

In [46]:
city_name = input("Введите название города: ")
current_temp = get_current_temperature(city_name)[1]
month = get_current_temperature(city_name)[2]
is_temperature_normal(city_name, current_temp, month, data)

Введите название города: Dubai


'Температура в норме'

Температуры для городов Берлин, Каир и Дубай в норме.

__5. Попробуйте для получения текущей температуры использовать синхронные и асинхронные методы. Что здесь лучше использовать?__

Давайте попробуем замерить скорость для синхронного метода, т.е. попробуем запросить температуры для некоторого количества городов:

In [136]:
city_list = ['Moscow', 'Berlin', 'Cairo']
start_time = time.time()
for i in city_list:
    print(f'{i}: {get_current_temperature(i)[0]}')
end_time = time.time()
print(f"Время выполнения c распараллеливанием составляет: {end_time - start_time:.2f} секунд")

Moscow: Текущая температура в городе Moscow: 1.55°C
Berlin: Текущая температура в городе Berlin: 5.95°C
Cairo: Текущая температура в городе Cairo: 21.42°C
Время выполнения c распараллеливанием составляет: 0.22 секунд


Теперь давайте попробуем асинхронный метод:

In [137]:
nest_asyncio.apply()

api_key = "MY_API_KEY"

async def get_current_temperature(city_name):
    url = "https://api.openweathermap.org/data/2.5/weather"
    params = {"q": city_name,
              "appid": api_key,
              "units": "metric"}

    async with aiohttp.ClientSession() as session:
        try:
            async with session.get(url, params=params) as response:
                data = await response.json()
                temperature = data["main"]["temp"]
                time = data["sys"]["sunrise"]
                season = pd.to_datetime(time, unit='s').month
                return f"Текущая температура в городе {city_name}: {temperature}°C", temperature, season

        except aiohttp.ClientError as e:
            return f"Ошибка при выполнении запроса: {e}"
        except KeyError:
            return "Ошибка: не удается получить информацию о температуре"

In [138]:
async def main():
    city_list = ['Moscow', 'Berlin', 'Cairo']
    tasks = [get_current_temperature(city) for city in city_list]
    results = await asyncio.gather(*tasks)
    for result in results:
        print(result[0])

start_time = time.time()
await main()
end_time = time.time()
print(f"Время выполнения c распараллеливанием составляет: {end_time - start_time:.2f} секунд")

Текущая температура в городе Moscow: 1.55°C
Текущая температура в городе Berlin: 5.95°C
Текущая температура в городе Cairo: 21.42°C
Время выполнения c распараллеливанием составляет: 0.08 секунд


Как результат: 0.22 секунд ссинхронного метода против 0.08 секунд асинхронного.Полагаю, что, если имеется нужда получать инфорамацию о погоде для списка городов, то лучшим вариантом будет использование асинхронных методов.

## Часть 3. Разработать интерактивное приложение

__Код для пунктов части 3 (см. ниже), можно найти в файле app.py.__
- Добавить интерфейс для загрузки файла с историческими данными.
- Добавить интерфейс для выбора города (из выпадающего списка).
- Добавить форму для ввода API-ключа OpenWeatherMap. Когда он не введен, данные для текущей погоды не показываются. Если ключ некорректный, выведите на экран ошибку (должно приходить {"cod":401, "message": "Invalid API key. Please see https://openweathermap.org/faq#error401 for more info."}).
- Отобразить:
1. Описательную статистику по историческим данным для города, можно добавить визуализации.
2. Временной ряд температур с выделением аномалий (например, точками другого цвета).
3. Сезонные профили с указанием среднего и стандартного отклонения.
4. Вывести текущую температуру через API и указать, нормальна ли она для сезона.